sdk/python/kfp/_local_client.py

```diff
@@ -0,0 +1,503 @@
+                    group_dag.add_edge(op.name, follow)
+                else:
+                    _base_group = self._find_base_group(group.groups, follow)
+                    if _base_group:
+                        # add edge to direct subgroup
+                        group_dag.add_edge(op.name, _base_group.name)
+
+            for dependency in pipeline_dag.get_dependencies(op.name):
+                if self._get_op(group.ops, dependency) is None:
+                    _base_group = self._find_base_group(group.groups, dependency)
+                    if _base_group:
+                        # add edge from direct subgroup
+                        group_dag.add_edge(_base_group.name, op.name)
+
+        return group_dag
+
+    def _create_op_dag(self, p: dsl.Pipeline) -> _Dag:
+        """ Create the DAG of the pipeline ops. """
+        dag = _Dag(p.ops.keys())
+
+        for op in p.ops.values():
+            # dependencies defined by inputs
+            for input_value in op.inputs:
+                if isinstance(input_value, dsl.PipelineParam):
+                    input_param = self._extract_pipeline_param(input_value.pattern)
+                    if input_param.op_name:
+                        dag.add_edge(input_param.op_name, op.name)
+                    else:
+                        logging.debug("%s depend on pipeline param", op.name)
+
+            # explicit dependencies of current op
+            for dependent in op.dependent_names:
+                dag.add_edge(dependent, op.name)
+        return dag
+
+    def _alter_output_file_path(
+        self, run_name: str, op_name: str, output_file: str
+    ) -> str:
+        """Alter the file path of output artifact to make sure it's unique in local runner.
+
+        kfp compiler will bound a tmp file for each component output, which is unique
+        in kfp runtime, but not unique in local runner. We alter the file path of the
+        name of current run and op, to make it unique in local runner.
+        """
+        return re.sub(
+            "/tmp",
+            "/{artifact_root}/{run_name}/{op_name}".format(
+                artifact_root=self._artifact_root,
+                run_name=run_name,
+                op_name=op_name.lower(),
+            ),
+            output_file,
+        )
+
+    def get_output_file_path(
+        self,
+        run_name: str,
+        pipeline: dsl.Pipeline,
+        op_name: str,
+        output_name: str = None,
+    ) -> str:
+        """ Get the file path of component output. """
+
+        op_dependency = pipeline.ops[op_name]
+        if output_name is None and len(op_dependency.file_outputs) == 1:
+            output_name = next(iter(op_dependency.file_outputs.keys()))
+        output_file = op_dependency.file_outputs[output_name]
+        altered_output_file = self._alter_output_file_path(
+            run_name, op_name, output_file
+        )
+        return altered_output_file
+
+    def _op_cmd_locally(
```

Comment:
Maybe rename to something like `_generate_cmd_for_subprocess_execution`
And consider renaming the sibling to `_generate_cmd_for_docker_execution`